File: cpp/source/communication/ControlData.cpp
#include "ControlData.hpp"

#include <string.h>

#ifdef __SKYDIVE_USE_STL__

#include <iomanip>
#include <sstream>

#include "SkyException.hpp"

#endif //__SKYDIVE_USE_STL__

ControlData::ControlData(void)
{
    euler = Vect3Dd();
    throttle = 0.0f;
}

ControlData::ControlData(const unsigned char* src)
{
    memcpy((unsigned char*)this + 4, src, getDataSize());
}

IMessage::PreambleType ControlData::getPreambleType(void) const
{
    return IMessage::CONTROL;
}

void ControlData::serialize(unsigned char* dst) const
{
    memcpy(dst, (unsigned char*)this + 4, getDataSize());
}

IMessage::MessageType ControlData::getMessageType(void) const
{
    return CONTROL_DATA;
}

unsigned ControlData::getDataSize(void) const
{
    return sizeof(ControlData) - 4;
}

void ControlData::setEuler(const Vect3Df& _euler)
{
    euler = _euler;
}

void ControlData::setThrottle(const float _throttle)
{
    throttle = _throttle;
}

void ControlData::setControllerCommand(const ControllerCommand& _controllerCommand)
{
    controllerCommand = static_cast<unsigned short>(_controllerCommand);
}

void ControlData::setSolverMode(const SolverMode& _solverMode)
{
    solverMode = static_cast<unsigned char>(_solverMode);
}

const Vect3Df& ControlData::getEuler(void) const
{
    return euler;
}

float ControlData::getThrottle(void) const
{
    return throttle;
}

ControlData::ControllerCommand ControlData::getControllerCommand(void) const
{
    return static_cast<ControllerCommand>(controllerCommand);
}

ControlData::SolverMode ControlData::getSolverMode(void) const
{
    return static_cast<SolverMode>(solverMode);
}

void ControlData::setRoll(const float roll)
{
    euler.x = roll;
}

void ControlData::setPitch(const float pitch)
{
    euler.y = pitch;
}

void ControlData::setYaw(const float yaw)
{
    euler.z = yaw;
}

#ifdef __SKYDIVE_USE_STL__

std::string ControlData::getControllerCommandString(const ControllerCommand& controllerCommand)
{
    switch (controllerCommand)
    {
    case MANUAL:
        return std::string("Manual");
    case AUTOLANDING:
        return std::string("Autolanding");
    case AUTOLANDING_AP:
        return std::string("Autolanding AP");
    case HOLD_ALTITUDE:
        return std::string("Hold: altitude");
    case HOLD_POSITION:
        return std::string("Hold: position");
    case BACK_TO_BASE:
        return std::string("Back to base");
    case VIA_ROUTE:
        return std::string("Fly via route");
    case STOP:
        return std::string("STOP");
    case ERROR_JOYSTICK:
        return std::string("ERROR Joystick");
    default:
        return std::string("ERROR");
    }
}

ControlData::ControllerCommand ControlData::getCommandFromString(const std::string& str)
{
    if (str == getControllerCommandString(ControlData::MANUAL))
    {
        return ControlData::MANUAL;
    }
    else if (str == getControllerCommandString(ControlData::AUTOLANDING))
    {
        return ControlData::AUTOLANDING;
    }
    else if (str == getControllerCommandString(ControlData::AUTOLANDING_AP))
    {
        return ControlData::AUTOLANDING_AP;
    }
    else if (str == getControllerCommandString(ControlData::HOLD_ALTITUDE))
    {
        return ControlData::HOLD_ALTITUDE;
    }
    else if (str == getControllerCommandString(ControlData::HOLD_POSITION))
    {
        return ControlData::HOLD_POSITION;
    }
    else if (str == getControllerCommandString(ControlData::BACK_TO_BASE))
    {
        return ControlData::BACK_TO_BASE;
    }
    else if (str == getControllerCommandString(ControlData::VIA_ROUTE))
    {
        return ControlData::VIA_ROUTE;
    }
    else if (str == getControllerCommandString(ControlData::STOP))
    {
        return ControlData::STOP;
    }
    else if (str == getControllerCommandString(ControlData::ERROR_JOYSTICK))
    {
        return ControlData::ERROR_JOYSTICK;
    }
    else
    {
        __SKY_EXCEPTION__("Unexpected command string");
    }
}

std::string ControlData::getSolverModeString(const SolverMode& solverMode)
{
    switch (solverMode)
    {
    case STABLILIZATION:
        return std::string("Stabilization");
    case ANGLE_NO_YAW:
        return std::string("Angle no yaw");
    case ANGLE:
        return std::string("Angle");
    case HEADLESS:
        return std::string("Headless");
    default:
        return std::string("Error");
    }
}

ControlData::SolverMode ControlData::getSolverModeFromString(const std::string& str)
{
    if (str == getSolverModeString(ControlData::STABLILIZATION))
    {
        return ControlData::STABLILIZATION;
    }
    else if (str == getSolverModeString(ControlData::ANGLE_NO_YAW))
    {
        return ControlData::ANGLE_NO_YAW;
    }
    else if (str == getSolverModeString(ControlData::ANGLE))
    {
        return ControlData::ANGLE;
    }
    else if (str == getSolverModeString(ControlData::HEADLESS))
    {
        return ControlData::HEADLESS;
    }
    else
    {
        __SKY_EXCEPTION__("Unexpected command string");
    }
}

ControlData ControlData::parseFromString(const std::string& line)
{
    ControlData controlData;
    std::istringstream iss(line);
    float time;
    int solverMode;
    if (!(iss
          >> time
          >> controlData.euler.x >> controlData.euler.y >> controlData.euler.z
          >> controlData.throttle
          >> controlData.controllerCommand
          >> solverMode
          ))
    {
        __SKY_EXCEPTION__("Error while parsing data.");
    }
    controlData.solverMode = (unsigned char)solverMode;
    return controlData;
}

std::ostream& operator << (std::ostream& stream, const ControlData& controlData)
{
    stream << controlData.euler.x << " " << controlData.euler.y << " " << controlData.euler.z << " ";
    stream << controlData.throttle << " ";
    stream << controlData.controllerCommand << " ";
    stream << (int)controlData.solverMode << " ";
    return stream;
}

#endif //__SKYDIVE_USE_STL__

File: java/data/ControlData.java
package com.skydive.java.data;

import com.skydive.java.CommMessage;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * Created by Bartosz Nawrot on 2016-08-19.
 * Main control message, it is send to controller synchronous with frequency of 20Hz
 * Contains basic control parameters:
 * - euler angles (roll, pitch, yaw) - desired rotation of UAV [rad]
 * - throttle - desired power of motors [0 - 1.0]
 * - command - command for specific control type
 * - solver mode - (only in MANUAL command), defines way of euler angles interpretation
 */
public class ControlData {

    private float roll, pitch, yaw;
    private float throttle;
    private ControllerCommand command;
    private SolverMode mode;

    public ControlData() {
        this.roll = 0.0f;
        this.pitch = 0.0f;
        this.yaw = 0.0f;
        this.throttle = 0.0f;
        this.command = ControllerCommand.MANUAL;
        this.mode = SolverMode.ANGLE_NO_YAW;
    }

    public ControlData(ControlData controlData) {
        this.roll = controlData.roll;
        this.pitch = controlData.pitch;
        this.yaw = controlData.yaw;
        this.throttle = controlData.throttle;
        this.command = controlData.command;
        this.mode = controlData.mode;
    }

    public ControlData(CommMessage message) {
        this(message.getPayload());
    }

    public ControlData(byte[] data) {
        this(data, 0);
    }

    public ControlData(byte[] data, int offset) {
        ByteBuffer buffer = ByteBuffer.wrap(data, offset,
                CommMessage.getPayloadSizeByType(CommMessage.MessageType.CONTROL));
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        this.roll = buffer.getFloat();
        this.pitch = buffer.getFloat();
        this.yaw = buffer.getFloat();
        this.throttle = buffer.getFloat();
        this.command = ControllerCommand.getControllerCommand(buffer.getShort());
        this.mode = SolverMode.getSolverMode(buffer.get());
    }

    public float getRoll() {
        return roll;
    }

    public void setRoll(float roll) {
        this.roll = roll;
    }

    public float getPitch() {
        return pitch;
    }

    public void setPitch(float pitch) {
        this.pitch = pitch;
    }

    public float getYaw() {
        return yaw;
    }

    public void setYaw(float yaw) {
        this.yaw = yaw;
    }

    public float getThrottle() {
        return throttle;
    }

    public void setThrottle(float throttle) {
        this.throttle = throttle;
    }

    public ControllerCommand getCommand() {
        return command;
    }

    public void setCommand(ControllerCommand command) {
        this.command = command;
    }

    public SolverMode getMode() {
        return mode;
    }

    public void setMode(SolverMode mode) {
        this.mode = mode;
    }

    public void setStopCommand() {
        roll = 0.0f;
        pitch = 0.0f;
        yaw = 0.0f;
        throttle = 0.0f;
        command = ControllerCommand.STOP;
    }

    @Override
    public String toString() {
        return "ControlData:[ roll: " + String.valueOf(roll)
                + ", pitch: " + String.valueOf(pitch)
                + ", yaw: " + String.valueOf(yaw)
                + ", throttle: " + String.valueOf(throttle)
                + ", Controller command: " + command.toString() + "(" + String.valueOf(command.getValue()) + ")"
                + ", solver mode: " + mode.toString() + "(" + String.valueOf(mode.getValue()) + ")"
                + " ]";
    }

    public byte[] serialize() {
        byte[] out = new byte[CommMessage.getPayloadSizeByType(CommMessage.MessageType.CONTROL)];
        serialize(out);
        return out;
    }

    public void serialize(byte[] out) {
        if (out.length >= CommMessage.getPayloadSizeByType(CommMessage.MessageType.CONTROL)) {
            ByteBuffer buffer = ByteBuffer.wrap(out);
            buffer.order(ByteOrder.LITTLE_ENDIAN);
            buffer.putFloat(getRoll());
            buffer.putFloat(getPitch());
            buffer.putFloat(getYaw());
            buffer.putFloat(getThrottle());
            buffer.putShort(getCommand().getValue());
            buffer.put(getMode().getValue());
        }
    }

    public CommMessage getMessage() {
        return new CommMessage(CommMessage.MessageType.CONTROL, serialize());
    }

    public static int getSize() {
        return CommMessage.getPayloadSizeByType(CommMessage.MessageType.CONTROL);
    }

    public enum ControllerCommand {
        // manual control
        MANUAL((short) 1000),
        // auto lading with specified descend rate
        AUTOLANDING((short) 1100),
        // auto lading with specified descend rate and position hold
        AUTOLANDING_AP((short) 1200),
        // auto altitude hold, throttle value is now specifying descend/climb rate
        // th = 0 -> -v, th = 0.5 -> 0, th = 1.0 -> v
        HOLD_ALTITUDE((short) 1300),
        // auto position hold, (hold altitude enabled)
        HOLD_POSITION((short) 1400),
        // autonomous back to base, climb 10 meters above start, cruise to base and auto land with AP
        BACK_TO_BASE((short) 1500),
        // cruise via specific route and back to base
        VIA_ROUTE((short) 1600),
        // immediate STOP (even when fling)
        STOP((short) 2000),
        // error conditions
        ERROR_CONNECTION((short) 6100),
        ERROR_JOYSTICK((short) 6200),
        ERROR_EXTERNAL((short) 6300);

        private final short value;

        ControllerCommand(short value) {
            this.value = value;
        }

        public short getValue() {
            return value;
        }

        static public ControllerCommand getControllerCommand(short value) {
            if (value == MANUAL.getValue()) return MANUAL;
            else if (value == AUTOLANDING.getValue()) return AUTOLANDING;
            else if (value == AUTOLANDING_AP.getValue()) return AUTOLANDING_AP;
            else if (value == HOLD_ALTITUDE.getValue()) return HOLD_ALTITUDE;
            else if (value == HOLD_POSITION.getValue()) return HOLD_POSITION;
            else if (value == BACK_TO_BASE.getValue()) return BACK_TO_BASE;
            else if (value == VIA_ROUTE.getValue()) return VIA_ROUTE;
            else if (value == STOP.getValue()) return STOP;
            else if (value == ERROR_CONNECTION.getValue()) return ERROR_CONNECTION;
            else if (value == ERROR_JOYSTICK.getValue()) return ERROR_JOYSTICK;
            else return ERROR_EXTERNAL; // TODO throw some exception
        }
    }

    public enum SolverMode {
        STABILIZATION((byte) 0),
        ANGLE_NO_YAW((byte) 1),
        ANGLE((byte) 2),
        HEADLESS((byte) 3);

        private byte value;

        SolverMode(byte value) {
            this.value = value;
        }

        public byte getValue() {
            return value;
        }

        static public SolverMode getSolverMode(byte value) {
            if (value == STABILIZATION.getValue()) return STABILIZATION;
            else if (value == ANGLE_NO_YAW.getValue()) return ANGLE_NO_YAW;
            else if (value == ANGLE.getValue()) return ANGLE;
            else if (value == HEADLESS.getValue()) return HEADLESS;
            else return ANGLE_NO_YAW; // TODO throw some exception
        }
    }

}
